### ISSUE TITLE
SQL: Introduce index names

### ISSUE THREAD

**Issue by devozerov** [hazelcast/hazelcast#15265]
Currently, `IMap` indexes are unnamed. This might be a problem from a user-convenience perspective. We should add a name to the index which will be displayed in metadata in the future:
* Add name getter/setter, make sure it is serialized
* If the name is not set, use some generated stuff (e.g. field names + unique ID)
* Check for name conflicts during the creation

**Comment by devozerov** [comment]
Motivation:
- When tuning performance users should be able to both create and drop indexes in runtime. In order to drop the index, the user should know its name
- SQL optimizer may pick the wrong access path sometimes. For this reason, database vendors have a hint to force the engine to use a specific index:
```sql
SELECT * FROM table /*+ INDEX my_idx +*/
```

**Comment by devozerov** [comment]
Note, that in addition to the name, our indexes miss other important capabilities:
- Attribute order (ASC/DESC)
- Current API cannot accommodate new index types (e.g. hash index) easily.
- In the future, we will create indexes on ReplicatedMap as well.
Taking this into count, the proposal is to slightly rework index configuration and index creation API and add an optional name attribute. If not provided, the index name will be auto-generated.
```java
// Index type enum is introduced
enum IndexType {
    ORDERED,
    HASH
}

// Renamed from MapIndexConfig because it will be used for ReplicatedMap as well
abstract class IndexConfig { 
    String name;

    abstract IndexType getType();
}

// Every index has a separate class, because different indexes may have different set of params
class SortedIndexConfig extends IndexConfig {
    String attribute;

    IndexType getType() {
        return IndexType.SORTED;
    }
}

class HashIndexConfig extends IndexConfig {
    String attribute;

    IndexType getType() {
        return IndexType.HASH;
    }
}

// addIndex now accepts config, as there will be many different potential index configurations
interface IMap {
    void addIndex(IndexConfig indexConfig);
}
```

**Comment by devozerov** [comment]
The aforementioned API was discussed with the team before publishing. Thanks to @mmedenjak for a proposal to have a different class per index type.

**Comment by devozerov** [comment]
PRs:
https://github.com/hazelcast/hazelcast/pull/15537
https://github.com/hazelcast/hazelcast-enterprise/pull/3154

**Comment by devozerov** [comment]
Implemented in 4.0.
